src/gui/theme/colortheme.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2017  Eugene Shalygin <eugene.shalygin@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef QBT_THEME_COLORTHEME_H
+#define QBT_THEME_COLORTHEME_H
+
+#include <QString>
+
+#include <memory>
+#include <map>
+
+class QColor;
+class QSettings;
+
+namespace BitTorrent
+{
+    enum class TorrentState;
+}
+
+namespace Log
+{
+    enum MsgType: int;
+}
+
+namespace Theme
+{
+    enum class DownloadProgressBarElement
+    {
+        Background,
+        Border,
+        Complete,
+        Incomplete
+    };
+
+    class ThemeInfo;
+
+    /*! \brief qBittorrent color theme
+     *
+     * Contains all colors that are needed to render a qBt UI, either using widgets or Web
+     */
+    class ColorTheme
+    {
+    public:
+        virtual ~ColorTheme();
+
+        virtual ThemeInfo info() const = 0;
+
+        virtual QColor torrentStateColor(const BitTorrent::TorrentState state) const = 0;
+        virtual QColor logMessageColor(const Log::MsgType messageType) const = 0;
+        virtual QColor downloadProgressBarColor(const DownloadProgressBarElement element) const = 0;
```

Comment:
Because different parts of the colours set require a bit different handling. Torrent state colours require fast access and as such are backed by a cache, while others do not need this. These three functions give a simple interface for clients and of course one can do:
```c++
class ColorTheme {
public:
    enum class Element {
       //...
       // long list
       // 
    };
protected:
   static Element mapToThemeElement(BitTorrent::TorrentState state);
   static Element mapToThemeElement(Log::MsgType messageType);
   static Element mapToThemeElement(DownloadProgressBarElement messageType);
   virtual QColor color (Element element, bool cached) const = 0;
   virtual void cacheElement(Element element);
public:
  QColor  color(BitTorrent::TorrentState state) const {
      return color(mapToElement(state), true);
  }
  QColor  color(Log::MsgType messageType) const {
      return color(mapToElement(state), false);
  }
};
```
> Then we won't need to change it every time we want to add support for those in some other application component.

Sorry, I can't see how adding an enum member is better or simpler than adding a virtual function.